crates/ruff_python_formatter/src/other/comprehension.rs

```diff
@@ -104,3 +105,17 @@ impl FormatNodeRule<Comprehension> for FormatComprehension {
         Ok(())
     }
 }
+
+struct ExprTupleWithoutParentheses<'a>(&'a Expr);
+
+impl Format<PyFormatContext<'_>> for ExprTupleWithoutParentheses<'_> {
+    fn fmt(&self, f: &mut Formatter<PyFormatContext<'_>>) -> FormatResult<()> {
+        match self.0 {
+            Expr::Tuple(expr_tuple) => expr_tuple
+                .format()
+                .with_options(TupleParentheses::Comprehension)
+                .fmt(f),
+            other => write!(f, [other.format()]),
```

Comment:
Should I be consistent here with a .fmt()? They're effectively the same thing, right?

Reply:
`write!` has slightly more overhead and it depends on the compiler if it is able to *reduce* it exactly to `other.format().fmt(f)`

The difference are:

* The `write!` macro wraps the `[...]` in an `Arguments` struct, and each entry in the array in an `Argument` (it calls the `format_args![]` macro internally
* `write` calls into `buffer.write_fmt` which iterates over all arguments https://github.com/astral-sh/ruff/blob/df9e64dba0f9e5f2dadf85c543bfa548dabd42ae/crates/ruff_formatter/src/lib.rs#L717-L724
* `write` accepts any object that implements `Buffer` as the first argument. So you can do
  ```rust
	let mut recording = f.start_recording();
	write!(recording, [...])?;
	let recorded = recording.stop();
  ```
  whereas `.fmt(f)` directly uses the passed `Formatter`.


I tend to use `.fmt` for single items, and `write![]` for sequences. The downside of this is that adding a new item requires changing the code from `fmt` to `write` (or the other way round) and I don't do this consistently. That's why I'm okay leaving this to individuals to decide what they prefer.

Reply:
48af71ad23e71b9d6cd6e40c0526cb5ef1fadc01